### PR TITLE
[LIVE-5076][LIVE-5081][LIVE-5083] Feature - Domain Support in signTransaction

### PR DESCRIPTION
**Pull request by lambertkevin**
### 📝 Description

This PR adds support for domains in transaction signatures. 

🎏  Feature flag is: **feature_domain_input_resolution** 
```json
  {
    "enabled": false,
    "params": {
      "supportedCurrencyIds": ["ethereum"]
    }
  }
```

### ❓ Context

- **Impacted projects**: `ledger-live-desktop`, `ledger-live-mobile`, `live-common`, `domain-service`, `hw-app-eth` 
- **Linked resource(s)**: 
- https://ledgerhq.atlassian.net/browse/LIVE-5076
- https://ledgerhq.atlassian.net/browse/LIVE-5081
- https://ledgerhq.atlassian.net/browse/LIVE-5083
### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/domain-service/src/resolvers/index.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ export const resolveDomain = async (
 
   const responses = Promise.allSettled(
     registries.map((registry) =>
-      network<string>({
+      axios.request<string>({
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/domain-service/src/resolvers/index.ts:34]: Because using the `network` module that is coming from `live-common` (for now, it could be changed once it becomes its own lib) so it's creating a cycle dependency once `@ledgerhq/domain-service` is added to `hw-app-eth`:

`@ledgerhq/live-common` —dep—> `hw-app-eth` —dep—> `@ledgerhq/live-common`

File: libs/ledgerjs/packages/hw-app-eth/src/Eth.ts [hunk 2]
```diff
@@ -1252,106 +1307,138 @@ export default class Eth {
...
+   * @param signature signature for the plugin
+   * @returns a boolean
+   */
+  setExternalPlugin(payload: string, signature: string): Promise<boolean> {
+    const payloadBuffer = Buffer.from(payload, "hex");
+    const signatureBuffer = Buffer.from(signature, "hex");
+    const buffer = Buffer.concat([payloadBuffer, signatureBuffer]);
+    return this.transport.send(0xe0, 0x12, 0x00, 0x00, buffer).then(
+      () => true,
+      (e) => {
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/src/Eth.ts:1344]: Actually the whole app binding is always using this `return this.transport` way except for sending chunks which requires multiple calls.

File: libs/ledgerjs/packages/hw-app-eth/src/modules/Domains/index.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import type { DomainServiceResolution as DomainDescriptor } from "@ledgerhq/domain-service/types";
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/src/modules/Domains/index.ts:1]: Yes it's on purpose but it's just a personal preference for modules & components (whatever is looking like a namespace I guess). It was already done this way for EIP712, but I can change it to lowercase if you prefer, I don't care that much about it 👌😛

File: libs/domain-service/src/__tests__/unit/resolvers.unit.test.ts [hunk 4]
```diff
@@ -1,33 +1,45 @@
-import network from "@ledgerhq/live-common/network";
```
Comment (hedi-edelbloute) [inline libs/domain-service/src/__tests__/unit/resolvers.unit.test.ts:1]: Why do we now use axios instead of network ?

File: libs/domain-service/src/signers/index.ts [hunk 5]
```diff
@@ -16,6 +17,11 @@ export const signDomainResolution = async (
   registryName: SupportedRegistries,
   challenge: string
 ): Promise<string | null> => {
+  if (!validateDomain(domain)) {
```
Comment (hedi-edelbloute) [inline libs/domain-service/src/signers/index.ts:20]: This change should imply unit tests updates

File: libs/domain-service/src/__tests__/functionnal/domain-service.react.test.tsx [hunk 6]
```diff
@@ -77,6 +77,7 @@ describe("useNamingService", () => {
         address: "forced mocked address",
```
Comment (hedi-edelbloute) [inline libs/domain-service/src/__tests__/functionnal/domain-service.react.test.tsx:77]: Not related to your changes but I think there is a typo on the folder, it's "functional" with a single n :D

**Review by hedi-edelbloute — APPROVED** [review]
lgtm

**Review by haammar-ledger — APPROVED** [review]
Nothing blocking, LGTM

Reviewing this PR makes it even clearer to me that we need all of these implementation docs :slightly_smiling_face: 
- ethereum VS evm families - differences and/or migration guide
- the concept of modules and resolution, specific to and only used in these families
- hw-app-eth

File: libs/domain-service/src/__tests__/unit/resolvers.unit.test.ts (hunk 4 above)
Comment (lambertkevin) [reply libs/domain-service/src/__tests__/unit/resolvers.unit.test.ts:1]: Same reason explained here: https://github.com/LedgerHQ/ledger-live/pull/3059#discussion_r1149021802

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/RecipientField.react.test.tsx [hunk 7]
```diff
@@ -1,19 +1,18 @@
...
 import BigNumber from "bignumber.js";
 import {
-  findCryptoCurrencyById,
+  getCryptoCurrencyById,
   setSupportedCurrencies,
 } from "@ledgerhq/live-common/currencies/index";
 import { Account } from "@ledgerhq/types-live";
 import { InvalidAddress } from "@ledgerhq/errors";
 import { ThemeProvider } from "styled-components";
-import network from "@ledgerhq/live-common/network";
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/RecipientField.react.test.tsx:11]: Yup, it's in a test file where I need to mock the API return, so I had to import Axios as dev dep and mock the lib to avoid calling the real backend.

File: libs/ledger-live-common/src/api/Ethereum.ts [hunk 8]
```diff
@@ -302,7 +302,7 @@ export const apiForCurrency = (currency: CryptoCurrency): API => {
         return data
           .map((m: any) => {
             if (!m || typeof m !== "object") return;
-            const { spender, value } = m;
+            const { spender, count: value } = m;
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledger-live-common/src/api/Ethereum.ts:305]: Not since v4... 
Not covered either, I'm just fixing it here because I was testing if I could remove some dead code from compound and realized no approval was ever fetched. That's the kind of things that would be catch by the integration tests I guess as we would see the missing approvals.
Not covering it here because our only use case for this method is compound which is (100% brokend and) being actively removed. 

FYI @hedi-edelbloute

File: libs/domain-service/src/tools/index.ts
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/domain-service/src/tools/index.ts:1]: I'll add a few comments, but to be honest that's a bit out of my (our) scope in general. This helper is just here to help debugging APDUs built by the backend. It's not actively used in the implementation at all. 
I'll ask ChatGPT to make me an ELI5 of what a TLV is ! 🤣

File: libs/ledgerjs/packages/hw-app-eth/tests/Eth.integration.test.ts [hunk 9]
```diff
@@ -367,8 +553,10 @@ test("provideERC20TokenInformation", async () => {
   const zrxInfo = byContractAddressAndChainId(
     "0xe41d2489571d322189246dafa5ebde1f4699f498",
     1
+  )!;
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/tests/Eth.integration.test.ts:556]: Since it's in a test file, I think the "t'inquiète 😉" is ok haha
If at some point the token doesn't exist, it's gonna be catched as an error either way while trying to get `zrxInfo.data` just after that and the test will fail. 
Other option would be to add a condition to check `zrxInfo` and throw if the token doesn't exist. I can change it this way if you prefer 👌

File: libs/ledgerjs/packages/hw-app-eth/src/services/types.ts [hunk 10]
```diff
@@ -1,3 +1,9 @@
+import { DomainServiceResolution as DomainDescriptor } from "@ledgerhq/domain-service/types";
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/src/services/types.ts:1]: The types were all intially set in `@ledgerhq/domain-service`, but when added as a dep to LLD/LLM, we were having a cycle dependency somewhere (partially due to the `network` module of LLC I guess). So moving the types to  `@ledgerhq/types-live` was a solution to have the right types in LLM/LLD and more specifically in test files for mocks. 

I'll try to move them again in domain-service now that we moved to `axios` and see if it's possible !

I'll add a few comments along the way 👍

File: libs/domain-service/src/tools/index.ts
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/domain-service/src/tools/index.ts:1]: > This helper is just here to help debugging APDUs built by the backend. It's not actively used in the implementation at all.

Sounds like a good start :slightly_smiling_face:

File: libs/ledgerjs/packages/hw-app-eth/tests/Eth.integration.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledgerjs/packages/hw-app-eth/tests/Eth.integration.test.ts:556]: If you ask me I can't think of a legit case to use this `!` operator, I would always prefer a safe check.
But definitely not a big deal in this case, it can stay that way up to you.

File: libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts [hunk 11]
```diff
@@ -52,6 +52,7 @@ describe("Ledger Service", () => {
         );
 
         expect(resolution).toEqual({
+          domains: [],
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts:55]: `domains` are not really handled by the `ledgerService` per say, it's just a passthrough to keep the logic of "_everything for clear signing should pass by the `ledgerService`_". 
I'm not a big fan of that, but that's the best solution I got to avoid a very poor DX where people would need to manually add things to the "resolution" after the `resolveTransaction` method was called. 

```typescript 
import Eth, { ledgerService } from "@ledgerhq/hw-app-eth"

const eth = new Eth(transport);
const tx = "e8018504e3b292008252089428ee52a8f3d6e5d15f8b131996950d7f296c7952872bd72a2487400080";
// ↓ this has been provided by the `@ledgerhq/domain-service` resolver at the beginning of the signature flow
const myDomain = { registry: "ens", address: "0x123", domain: "coucou.eth", type: "forward" };
const resolution = await ledgerService.resolveTransaction(tx, null, { erc20: true, nft: true }); 
// ↑ this is supposed to handle the whole clear signing shenanigans
  
// kinda weird to have to do that here
if (youGotADomainToClearSign) {
  resolution.domains = [myDomain];
}

const result = eth.signTransaction("44'/60'/0'/0/0", tx, resolution); 
// Or we would have to add it here as a new parameter in the signTransaction method, 
// which makes it less and less a binding to the nano app + it creates a new signature that 
// might bites us in the ass later if we need yet another parameter
```

VS 

```typescript 
import Eth, { ledgerService } from "@ledgerhq/hw-app-eth"

const eth = new Eth(transport);
const tx = "e8018504e3b292008252089428ee52a8f3d6e5d15f8b131996950d7f296c7952872bd72a2487400080";
// ↓ this has been provided by the `@ledgerhq/domain-service` resolver at the beginning of the signature flow
const myDomain = { registry: "ens", address: "0x123", domain: "coucou.eth", type: "forward" };
const resolution = await ledgerService.resolveTransaction(tx, null, { erc20: true, nft: true, domains: [myDomain] }); 

const result = eth.signTransaction("44'/60'/0'/0/0", tx, resolution); 
```

So to answer the question, no there is no need for new tests there, as `ledgerService.resolveTransaction` is doing nothing with the domains but to keep them as is in the `resolution` variable to make sure `eth.signTransaction` as access to it.

**SIDE NOTE:**
Also why can't `ledgerService.resolveTransaction` get the APDUs itself like with ERC20/NFTs ? 
→ Because it requires a "challenge" provided by the nano app. So the other option would be:

```typescript
import Eth, { ledgerService } from "@ledgerhq/hw-app-eth"

const eth = new Eth(transport);
const tx = "e8018504e3b292008252089428ee52a8f3d6e5d15f8b131996950d7f296c7952872bd72a2487400080";
const myDomain = { registry: "ens", address: "0x123", domain: "coucou.eth", type: "forward" };
const challenge = myDomain ? eth.getChallenge() : undefined; // Requiring the nano to be active already
const resolution = await ledgerService.resolveTransaction(
  tx, 
  null,
  { 
    erc20: true, 
    nft: true, 
    domains: [myDomain], 
    challenge // necessary for sending the domain APDU to the nano app
  }
); 

const result = eth.signTransaction("44'/60'/0'/0/0", tx, resolution); 
```

This means that in our own logic in the `signOperation` Observable, you would already need the nano before the step `device-signature-requested`, which is against what :gre: wants 🤣

[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts:55]: Neat explanation :ok_hand: thanks man.
Would be very cool in a dedicated doc somewhere... :stuck_out_tongue:

File: libs/ledgerjs/packages/hw-app-eth/src/services/types.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/src/services/types.ts:1]: Confirmed, not possible because the type `Transaction` is now using the `DomainServiceResolution`, and adding `@ledgerhq/domain-service` as a dev dep to `@ledgerhq/types-live` is creating a cycle dependency when build live-common

[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledgerjs/packages/hw-app-eth/src/services/types.ts:1]: This suggests that we need a dedicated `@ledger/types-domain` lib, I think that would be the cleanest pattern, but NBD for now though, just a possible improvement.

File: libs/domain-service/src/signers/index.ts (hunk 5 above)
Comment (lambertkevin) [reply libs/domain-service/src/signers/index.ts:20]: Test added but very redundant with the already existing tests for `validateDomain` itself.

File: libs/domain-service/src/tools/index.ts
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/domain-service/src/tools/index.ts:1]: Done 👌

File: libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts:55]: Well that 100% the kind of things that would go into the "Explanation" part of a [diataxis](https://diataxis.fr/) documentation for example... If you see what I mean... 😏 :neckbeard:

File: libs/ledgerjs/packages/hw-app-eth/src/services/types.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledgerjs/packages/hw-app-eth/src/services/types.ts:1]: To be honest I'm really not sure having "types" libs for everything is a nice pattern 😕

File: libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledgerjs/packages/hw-app-eth/tests/ledgerService.unit.test.ts:55]: _touché_ :slightly_smiling_face: we're working on the topic with @lvndry , but until then good ol' Confluence is better than nothing.

File: libs/ledgerjs/packages/hw-app-eth/src/services/types.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledgerjs/packages/hw-app-eth/src/services/types.ts:1]: Not everything but seems to make sense for libs that can be used in more than 1 project.
It's a solution to circular dependencies :man_shrugging: seems better than split types.
What do you think is the issue? Too many libs everywhere? Or something worse?
